rfcs/rfc-007-dynamic-program-dispatch.md

```diff
@@ -0,0 +1,183 @@
+
+This document highlights a set of changes to enable a framework for Solana
+programs to be composed and dynamically dispatched.
+
+### Account additions
+
+```diff
+diff --git a/common/src/account.rs b/common/src/account.rs
+index 5abd642..89290ea 100644
+--- a/common/src/account.rs
++++ b/common/src/account.rs
+@@ -11,6 +11,12 @@ pub struct Account {
+     pub userdata: Vec<u8>,
+     /// contract id this contract belongs to
+     pub program_id: Pubkey,
++
++    /// this account contains a program (and is strictly read-only)
++    pub executable: bool,
```

Comment:
Yeah that's explicitly enabled here: NativeLoader loads BpfLoader and SoluaLoader.  Conceivable we could then have a lua script that implements the `LoaderInstruction` interface for a Budget DSL-like language.

Reply:
But after about the 3 chained loader or so, I'd expect the language to devolve into something so abstract that the `LoaderInstruction` interface itself can no longer be expressed in that language and you've reached peak loader.

Reply:
Are loader's marked as executable?

Reply:
One related property I'd like to see, say I had two versions of solua, one that was loaded by NativeLoader, and one that was loaded by BpfLoader (that was of course loaded by NativeLoader). Can we ensure that colua programs can's tell the difference? Besides pointing to `colua_via_bpf_program_id` instead of `lua_program_id`. That is, the accounts it sees will be the same length and in the same order?

Reply:
If you tack on the KeyAccounts as N+1, N+2, you'd be able to see a difference. If you can pass in a slice, then it should be straightforward to ensure the loaded program has no access to the loader accounts.

Reply:
It's the same either way.  The extra accounts get peeled off the front or back before the loaded program gains control.  But with KeyAccounts, my concern about prepending is gone, so let's just do that